lightwood/encoders/text/pretrained.py

```diff
@@ -0,0 +1,323 @@
+from lightwood.logger import log
+from lightwood.helpers.torch import LightwoodAutocast
+
+from transformers import (
+    DistilBertModel,
+    DistilBertForSequenceClassification,
+    DistilBertTokenizerFast,
+    AdamW,
+    get_linear_schedule_with_warmup,
+)
+
+
+class PretrainedLang(BaseEncoder):
+    """
+    Pretrained language models.
+    Option to train on a target encoding of choice.
+
+    Args:
+    is_target ::Bool; data column is the target of ML.
+    model_name ::str; name of pre-trained model
+    max_training_time ::int; seconds to train
+    custom_tokenizer ::function; custom tokenizing function
+    batch_size  ::int; size of batch
+    max_position_embeddings ::int; max sequence length of input text
+    custom_train ::Bool; If true, trains model on target procided
+    frozen ::Bool; If true, freezes transformer layers during training.
+    epochs ::int; number of epochs to train model with
+    """
+
+    def __init__(
+        self,
+        is_target=False,
+        model_name="distilbert",
+        custom_tokenizer=None,
+        batch_size=10,
+        max_position_embeddings=None,
+        custom_train=True,
+        frozen=False,
+        epochs=1,
+    ):
+        super().__init__(is_target)
+
+        self.name = model_name + " text encoder"
+        print(self.name, flush=True)
+
+        self._max_len = max_position_embeddings
+        self._custom_train = custom_train
+        self._frozen = frozen
+        self._batch_size = batch_size
+        self._epochs = epochs
+
+        # Model setup
+        self._tokenizer = custom_tokenizer
+        self._model = None
+        self.model_type = None
+
+        # TODO: Other LMs; Distilbert is a good balance of speed/performance
+        self._classifier_model_class = DistilBertForSequenceClassification
+        self._embeddings_model_class = DistilBertModel
+        self._tokenizer_class = DistilBertTokenizerFast
+        self._pretrained_model_name = "distilbert-base-uncased"
+
+        self.device, _ = get_devices()
+
+    def prepare(self, priming_data, training_data=None):
+        """
+        Prepare the encoder by training on the target.
+
+        Training data must be a dict with "targets" avail.
+        Automatically assumes this.
+        """
+        if self._prepared:
+            raise Exception("Encoder is already prepared.", flush=True)
+
+        # TODO: Make tokenizer custom with partial function; feed custom->model
+        if self._tokenizer is None:
+            self._tokenizer = self._tokenizer_class.from_pretrained(
+                self._pretrained_model_name
+            )
+
+        # Replaces empty strings with ''
+        priming_data = [x if x is not None else "" for x in priming_data]
+
+        # Checks training data details
+        # TODO: The dict should be streamlined
+        output_avail = training_data is not None and len(training_data["targets"]) == 1
+
+        output_type = training_data["targets"][0]["output_type"]
```

Comment:
I know that making this work with multiple-target is really hard, but at the moment this will break the whole thing and is very bad.

Please check if the output length is == 1 and then do the training, otherwise use the raw embeddings. As it stands this risks breaking all multi-target models in weird ways.

Reply:
ok, can add

Reply:
doesn't "len(training_data["targets"]) == 1" take care of it?

Reply:
oh never mind - i understand your point